### ISSUE TITLE
key 'epoch' in model file is -1

### ISSUE THREAD

**Issue by hubutui** [ultralytics/yolov5#1715]
## ❔Question
I loaded the `best.pt` after training, and found the `epoch` is -1. However, according to https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/train.py#L371-L383
`epoch` is saved in checkpoint file too. What's wrong?
I also try the trained model you provide, but `epoch` is also -1.

I'm using commit 94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5

**Comment by glenn-jocher** [comment]
@hubutui pretrained models all display epoch=-1 to allow them to be used as pretrained models, where training starts at epoch 0.

**Comment by hubutui** [comment]
@glenn-jocher Yeah, but I check the model I train from scratch, it's `epoch` is also -1.

**Comment by glenn-jocher** [comment]
@hubutui yes, every model that finishes training is reset to epoch=-1 to allow it to be used as a pretrained model for finetuning future models. For example `python train.py --weights yolov5s.pt` uses a 300 epoch trained model for pretrained weights. 

If training is interrupted, last.pt will have epochs= the last epoch, and can be resumed from that epoch with python train.py --resume. Fully trained models are not resumable, as their LR scheduler has completed to near zero LR.

**Comment by hubutui** [comment]
@glenn-jocher Ok, I see. I was trying to see which epoch is the best in the validation set. So, I have to run the test on the validation set after training. However, the test result is not matched with the `results.txt` exactly. Do you use different config in the validation step in training and `test.py`?

**Comment by glenn-jocher** [comment]
@hubutui best.pt is saved at the maximum fitness epoch during training. You can view training logs such as results.txt or W&B dashboard to see maximum fitness epoch. Fitness is defined a weighted combination of metrics here:
https://github.com/ultralytics/yolov5/blob/49abc722fc4204d57e81b2f4db98f024ad6474cc/utils/metrics.py#L12-L16

**Comment by glenn-jocher** [comment]
@hubutui train mAP, and test mAP should be identical now, we had users raise issues in the past with slight divergences. This was updated in PR https://github.com/ultralytics/yolov5/pull/1616.
